rollup/ccc/async_checker.go

```diff
@@ -0,0 +1,176 @@
+package ccc
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/sourcegraph/conc/stream"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core"
+	"github.com/scroll-tech/go-ethereum/core/rawdb"
+	"github.com/scroll-tech/go-ethereum/core/state"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/core/vm"
+	"github.com/scroll-tech/go-ethereum/ethdb"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/metrics"
+	"github.com/scroll-tech/go-ethereum/params"
+	"github.com/scroll-tech/go-ethereum/rollup/tracing"
+)
+
+var (
+	failCounter        = metrics.NewRegisteredCounter("ccc/async/fail", nil)
+	checkTimer         = metrics.NewRegisteredTimer("ccc/async/check", nil)
+	activeWorkersGauge = metrics.NewRegisteredGauge("ccc/async/active_workers", nil)
+)
+
+type Blockchain interface {
+	Database() ethdb.Database
+	GetBlock(hash common.Hash, number uint64) *types.Block
+	StateAt(root common.Hash) (*state.StateDB, error)
+	Config() *params.ChainConfig
+	GetVMConfig() *vm.Config
+	core.ChainContext
+}
+
+// AsyncChecker allows a caller to spawn CCC verification tasks
+type AsyncChecker struct {
+	bc             Blockchain
+	onFailingBlock func(*types.Block, error)
+
+	workers        *stream.Stream
+	checkers       []*Checker
+	freeCheckerIdx int
+}
+
+type ErrorWithTxnIdx struct {
+	txIdx uint
+	err   error
+}
+
+func (e *ErrorWithTxnIdx) Error() string {
+	return fmt.Sprintf("txn at index %d failed with %s", e.txIdx, e.err)
+}
+
+func (e *ErrorWithTxnIdx) Unwrap() error {
+	return e.err
+}
+
+func NewAsyncChecker(bc Blockchain, numWorkers int) *AsyncChecker {
+	return &AsyncChecker{
+		bc: bc,
+		checkers: func(count int) []*Checker {
+			checkers := make([]*Checker, count)
+			for idx := range checkers {
+				checkers[idx] = NewChecker(true)
+			}
+			return checkers
+		}(numWorkers),
+		workers: stream.New().WithMaxGoroutines(numWorkers),
+	}
+}
+
+func (c *AsyncChecker) WithOnFailingBlock(onFailingBlock func(*types.Block, error)) *AsyncChecker {
+	c.onFailingBlock = onFailingBlock
+	return c
+}
+
+func (c *AsyncChecker) Wait() {
+	c.workers.Wait()
+}
+
+// Check spawns an async CCC verification task.
+func (c *AsyncChecker) Check(block *types.Block) error {
+	checker := c.checkers[c.freeCheckerIdx]
+	c.freeCheckerIdx = (c.freeCheckerIdx + 1) % len(c.checkers)
```

Comment:
if we have a lot of blocks, how do we ensure this checker is free?

Reply:
also, although it's extremely unlikely, but I feel like it's safer if we add a lock here, otherwise another `Check` might get the same `checker := c.checkers[c.freeCheckerIdx]` before a previous `Check` updates the `freeCheckerIdx`

Reply:
Check is supposed to be called by only a single caller, so there shouldn't be any concurrency issues. But I will replace the slice with a channel just to be safe.